README.md
# Stashward

[![Build Status](https://travis-ci.org/PSU-OIT-ARC/stashward.svg?branch=master)](https://travis-ci.org/PSU-OIT-ARC/stashward)

Stashward is a log formatter and handler for Python that implements the (poorly specified)
logstash-forwarder protocol.

The protocol is described in what *appears* to be good enough detail here:
https://github.com/elasticsearch/logstash-forwarder/blob/master/PROTOCOL.md

But if you implement the protocol as described, you will crash logstash every time
you send a data packet to it.

The trick is that there are three **required** key/value pairs: host, line and
offset. Those must be sent with every packet. And you also **cannot** send an
`@timestamp` field. I do not know if that is a logstash-forwarder issue, or a
logstash issue.

# Install

    pip install stashward

# Security

Following logstash-forwarder's lead, a server SSL certificate is required, and
it must be signed by a CA. The common name on the certificate must match the
host you are connecting to.

# Usage

```python
import logging
from stashward import StashwardHandler

root = logging.getLogger('')
root.setLevel(logging.INFO)

# give it a host, and port where logstash forwarder is listening, and the path
# to a CA file
handler = StashwardHandler("example.com", 5043, ca_certs="/etc/pki/tls/certs/ALL_CAs.crt")
root.addHandler(handler)

# log something
logging.info("Foo")

# log something with an exception
try:
    raise ValueError("foo")
except ValueError:
    logging.exception("Foobar!")
```

# Example Logstash Config

```
input {
    lumberjack {
        port => 5043
        type => "lumberjack"
        ssl_key => "/etc/logstash/private.key"
        ssl_certificate => "/etc/logstash/a_CA_signed_certificate_file.crt"
    }
}

output {
    elasticsearch {
        host => localhost
        protocol => http
    }
}
```

# Testing

The unit tests can be run in Python2 and Python3 by running:

    make test


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.3] - 2015-04-28
### Fixed
- A mysterious 0.1.2 version was already on PyPI, so I am bumping the version
  number

## [0.1.2] - 2015-04-27
### Added
- Python 2 compatibility when unicode (text) data is emitted.

## [0.1.1] - 2015-03-11
### Added
- Everything


runtests.py
#!/usr/bin/env python
import logging
import os
import socket
import sys
import tempfile
import time
try:
    import unittest2 as unittest
except ImportError:
    import unittest
from ssl import wrap_socket
from multiprocessing import Process
from stashward import StashwardHandler
from stashward.handler import ErrorCode


class FullTest(unittest.TestCase):
    """
    This test creates a listening SSL socket on an abitrary port and waits for
    connections on another process. We then setup the log handler that connects
    on that socket, and send logging messages to it.
    """

    def setUp(self):
        super(FullTest, self).setUp()
        # create the socket, and bind to an abitrary port
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s = wrap_socket(s, keyfile="certs/localhost.key", certfile="certs/localhost.crt", server_side=True)
        s.bind(("localhost", 0))
        s.listen(1)
        self.port = s.getsockname()[1]

        def ensure_messages_are_correct(socket):
            """
            Wait for the client to connect, and then ensure the messages
            received and are correct
            """
            # wait until someone connects...
            client, address = socket.accept()
            # the first message should be the window size, that starts with 1D, and
            # ends with 2**32-1 in big endian format
            assert(client.read() == b'1W\xff\xff\xff\xff')
            # the second message should start with the 1D packet header, the
            # sequence number (0), and the number of key/value pairs, which is
            # 9 (@version, message, levelname, logger, etc)
            assert(client.read()[:10] == b'1D\x00\x00\x00\x00\x00\x00\x00\x09')
            # the third message should contain strack trace information
            assert(b"Traceback" in client.read())
            s.close()

        # start the listening process
        Process(target=ensure_messages_are_correct, args=[s]).start()

    def test(self):
        """
        Ensure the logger gets the 3 messages: One for the window size, and two
        data packets
        """
        root = logging.getLogger('')
        root.setLevel(logging.INFO)
        handler = StashwardHandler("localhost", self.port, ca_certs="certs/localhost.crt")
        root.addHandler(handler)
        logging.info("Foo")
        try:
            raise ValueError("foo")
        except ValueError:
            logging.exception("Foobar!")
        handler.close()


class TestHandler(unittest.TestCase):
    def test_makePickle(self):
        """Ensure the sequence number rolls over"""
        handler = StashwardHandler("google.com", 443, ca_certs="certs/ca-bundle.crt")
        handler.sequence_number = 2**32-1

        # mock up a log record
        class Record(dict):
            def getMessage(self):
                return ''

            def __getattr__(self, key):
                return ''

        handler.makePickle(Record())
        self.assertEqual(handler.sequence_number, 0)
        handler.close()

    def test_packetize(self):
        """Ensure the generated packet follows the right protocol"""
        handler = StashwardHandler("google.com", 443, ca_certs="certs/ca-bundle.crt")
        # ensure these get converted to a packet with string keys and string values
        d = {"foo": "bar", 4: 4}
        arbitrary_sequence_number = 2503748420
        packet = handler.packetize(d, arbitrary_sequence_number)
        # we'll get a different looking package depending on the order of the dict
        if list(d.keys())[0] == "foo":
            self.assertEqual(packet, b'1D\x95\x3c\x2b\x44\x00\x00\x00\x02\x00\x00\x00\x03foo\x00\x00\x00\x03bar\x00\x00\x00\x014\x00\x00\x00\x014')
        else:
            self.assertEqual(packet, b'1D\x95\x3c\x2b\x44\x00\x00\x00\x02\x00\x00\x00\x014\x00\x00\x00\x014\x00\x00\x00\x03foo\x00\x00\x00\x03bar')
        handler.close()

    def test_unicode(self):
        """This is a regression test for handling unicode in Python2"""
        handler = StashwardHandler("google.com", 443, ca_certs="certs/ca-bundle.crt")
        # the unicode codepoint 044f (in hex) (or 1103 in decimal) is a
        # backwards looking R character. The UTF8 encoding of this particular
        # character is 0xD18F
        d = {u"\u044f": u"\u044f"}
        arbitrary_sequence_number = 5
        packet = handler.packetize(d, arbitrary_sequence_number)
        self.assertEqual(packet, b'1D\x00\x00\x00\x05\x00\x00\x00\x01\x00\x00\x00\x02\xd1\x8f\x00\x00\x00\x02\xd1\x8f')
        handler.close()


class TestSSL(unittest.TestCase):
    def setUp(self):
        super(TestSSL, self).setUp()
        self.stderr = tempfile.TemporaryFile("w+")

    def tearDown(self):
        self.stderr.close()
        super(TestSSL, self).tearDown()

    def test_ca_file_not_found(self):
        """
        Ensure a helpful error is printed to stderr if the ca file is not
        found, and the socket doesn't get created.
        """
        with RedirectStdStreams(stderr=self.stderr):
            handler = StashwardHandler("google.com", 443, ca_certs=".file_that_does_not_exist")
        self.stderr.seek(0)
        content = self.stderr.read()
        # Python3 gives better error messages
        if sys.version_info[0] >= 3:
            self.assertIn("ERROR %d:" % ErrorCode.FILE_NOT_FOUND, content)
            self.assertIn("FileNotFoundError", content)
        else:
            self.assertIn("ERROR %d:" % ErrorCode.GENERIC_SSL_FAILURE, content)
        # ensure the socket didn't get created somehow
        self.assertFalse(handler.sock)
        handler.close()

    def test_ca_file_is_invalid(self):
        """
        Ensure a helpful error is printed to stderr if the ca file is invalid
        and the socket doesn't get created
        """
        with RedirectStdStreams(stderr=self.stderr):
            # server.key is not not a valid CA file
            handler = StashwardHandler("google.com", 443, ca_certs="certs/localhost.key")
        self.stderr.seek(0)
        content = self.stderr.read()
        self.assertIn("ERROR %d:" % ErrorCode.GENERIC_SSL_FAILURE, content)
        # ensure the socket didn't get created somehow
        self.assertFalse(handler.sock)
        handler.close()

    def test_not_a_trusted_cert(self):
        """
        Ensure a helpful error is printed to stderr if the certificate is not
        trusted and the socket doesn't get created
        """
        with RedirectStdStreams(stderr=self.stderr):
            # Google isn't trusted by server.crt
            handler = StashwardHandler("google.com", 443, ca_certs="certs/localhost.crt")
        self.stderr.seek(0)
        content = self.stderr.read()
        self.assertIn("ERROR %d:" % ErrorCode.GENERIC_SSL_FAILURE, content)
        # check for the better error message in Python 3
        if sys.version_info[0] >= 3:
            self.assertIn("CERTIFICATE_VERIFY_FAILED", content)
        # ensure the socket didn't get created somehow
        self.assertFalse(handler.sock)
        handler.close()

    def test_cn_mismatch(self):
        """
        Ensure a helpful error is printed to stderr if the certificate is
        trusted by the CN is a mismatch and the socket doesn't get created
        """
        google_ip = socket.gethostbyname('google.com')
        with RedirectStdStreams(stderr=self.stderr):
            # The certificate for google isn't issued for the IP address
            handler = StashwardHandler(google_ip, 443, ca_certs="certs/ca-bundle.crt")
        self.stderr.seek(0)
        content = self.stderr.read()
        self.assertIn("ERROR %d:" % ErrorCode.CN_MISMATCH, content)
        # ensure the socket didn't get created somehow
        self.assertFalse(handler.sock)
        handler.close()


# Thanks http://stackoverflow.com/a/6796752/2733517
class RedirectStdStreams(object):
    def __init__(self, stdout=None, stderr=None):
        self._stdout = stdout or sys.stdout
        self._stderr = stderr or sys.stderr

    def __enter__(self):
        self.old_stdout, self.old_stderr = sys.stdout, sys.stderr
        self.old_stdout.flush(); self.old_stderr.flush()
        sys.stdout, sys.stderr = self._stdout, self._stderr

    def __exit__(self, exc_type, exc_value, traceback):
        self._stdout.flush(); self._stderr.flush()
        sys.stdout = self.old_stdout
        sys.stderr = self.old_stderr


if __name__ == '__main__':
    unittest.main()


stashward/handler.py
from __future__ import unicode_literals
from __future__ import print_function

import traceback
import struct
import logging
import errno
import socket
import sys
import six

from logging.handlers import SocketHandler
from ssl import wrap_socket, CERT_REQUIRED, SSLError

try:
    from ssl import match_hostname, CertificateError
except ImportError:
    from backports.ssl_match_hostname import match_hostname, CertificateError
from .formatter import StashwardFormatter


class ErrorCode:
    """
    We use some random numbers for error codes to make unit testing a little
    easier, and for greppability
    """
    GENERIC_SSL_FAILURE = 42218
    CN_MISMATCH = 34624
    FILE_NOT_FOUND = 17253

class StashwardHandler(SocketHandler):
    # The sequence number is 32 bits, so it should rollover to 0 when it gets there
    MAX_SEQUENCE_NUMBER = 2**32-1

    def __init__(self, host, port, ca_certs, message_type="django"):
        """
        Specify the host, port, and file path to the concatenated list of CAs
        """
        SocketHandler.__init__(self, host, port)
        self.formatter = StashwardFormatter(message_type)
        self.ca_certs = ca_certs
        self.sequence_number = 0

        # immediately send a packet setting the windows size to something huge
        # since we don't care about ACKs
        self.send(struct.pack(b"!ssI", b"1", b"W", self.MAX_SEQUENCE_NUMBER))

    def makeSocket(self):
        """Make the socket and wrap it with SSL. A valid certificate is required from a trusted CA"""
        s = SocketHandler.makeSocket(self)

        try:
            s = wrap_socket(s, cert_reqs=CERT_REQUIRED, ca_certs=self.ca_certs)
        except SSLError as e:
            # the ssl certificate was probably bad
            # since logging doesn't work, print to stderr
            traceback.print_exc()
            print("ERROR %d: This is likely caused by a CN mismatch, an invalid certificate file at %s, or a verification failure based on your CA file" % (ErrorCode.GENERIC_SSL_FAILURE, self.ca_certs), file=sys.stderr)
            s.close()
            raise
        except OSError as e:
            s.close()
            # if we get an ENOENT, that's because the ca_file is bad
            if e.errno == errno.ENOENT:
                traceback.print_exc()
                print("ERROR %d: This error is probably caused by a bad CA file at %s" % (ErrorCode.FILE_NOT_FOUND, self.ca_certs), file=sys.stderr)
            s.close()
            raise

        # some python versions do not do a CN match check, so we have to do it
        try:
            match_hostname(s.getpeercert(), self.host)
        except CertificateError as e:
            traceback.print_exc()
            print("ERROR %d: The CN provided by the server didn't match the host you connected to" % ErrorCode.CN_MISMATCH, file=sys.stderr)
            s.close()
            # raising a socket error will trigger createSocket to retry after a period of time
            raise socket.error("CN Mismatch")

        return s

    def makePickle(self, record):
        """Format the record, and turn it into a logstash compatible packet"""
        packet = self.packetize(self.formatter.format(record), self.sequence_number)
        # critical section here for incremeting the sequence_number. Notice, we
        # don't have to wrap a lock around this, because the base
        # logging.Handler will acquire a lock before calling emit(), which
        # calls this method
        self.sequence_number = self.sequence_number + 1 if self.sequence_number < self.MAX_SEQUENCE_NUMBER else 0
        return packet

    def packetize(self, data, sequence_number):
        """
        Turn a dict into a byte stream that is compatible with logstash-forwarder's protocol

        The packet looks like this:

        Header "1D" - the version and frame type (data) in ascii
        32 bit unsigned, big-endian, integer sequence number
        32 bit unsigned, big-endian, integer count of the number of (key, value) pairs
        For each key, value pair:
            32 bit unsigned, big-endian, integer number for the key length
            the key (raw bytes)
            32 bit unsigned, big-endian, integer number for the value length
            the value (raw bytes)


        The data *MUST* include a key/value pair for "host", "line" and
        "offset". That fact isn't documented anywhere but here. You're welcome.
        """
        format_string = [b"!ssII"]
        string = [b"1", b"D", sequence_number, len(data)]

        for key, value in data.items():
            # encode the keys and values as utf8-encoded bytestrings
            key = six.text_type(key).encode('utf-8')
            value = six.text_type(value).encode('utf-8')

            # tack on the key length
            format_string.append(b"I")
            string.append(len(key))
            # tack on the key itself
            format_string.append(six.text_type("%ds" % (len(key))).encode('utf-8'))
            string.append(key)
            # tack on the value length
            format_string.append(b"I")
            string.append(len(value))
            # tack on the value
            format_string.append(six.text_type("%ds" % (len(value))).encode('utf-8'))
            string.append(value)

        return struct.pack(b"".join(format_string), *string)


stashward/formatter.py
import logging
import socket
import sys


# Code shameless taken from Vladimir Klochan
# https://github.com/vklochan/python-logstash/blob/2f8b5e7c0befe08d0b39c3929c0a78dbad1bd014/logstash/formatter.py
class StashwardFormatter(logging.Formatter):
    def __init__(self, message_type="django"):
        self.host = socket.gethostname()
        self.message_type = message_type

    def get_extra_fields(self, record):
        """Tack on all the fields from the extra parameter on logging calls"""
        # The list contains all the attributes listed in
        # http://docs.python.org/library/logging.html#logrecord-attributes
        skip_list = set([
            'args', 'asctime', 'created', 'exc_info', 'exc_text', 'filename',
            'funcName', 'id', 'levelname', 'levelno', 'lineno', 'module',
            'msecs', 'msecs', 'message', 'msg', 'name', 'pathname', 'process',
            'processName', 'relativeCreated', 'thread', 'threadName', 'extra', 'stack_info'])

        if sys.version_info < (3, 0):
            easy_types = (basestring, bool, dict, float, int, list, type(None))
        else:
            easy_types = (str, bool, dict, float, int, list, type(None))

        fields = {}

        for key, value in record.__dict__.items():
            if key not in skip_list:
                if isinstance(value, easy_types):
                    fields[key] = value
                else:
                    fields[key] = repr(value)

        return fields

    def get_debug_fields(self, record):
        """Format the exception info"""
        fields = {
            'exc_info': self.formatException(record.exc_info),
            'process': record.process,
        }

        # funcName was added in 2.5
        if not getattr(record, 'funcName', None):
            fields['funcName'] = record.funcName

        # processName was added in 2.6
        if not getattr(record, 'processName', None):
            fields['processName'] = record.processName

        return fields

    def format(self, record):
        """Create a dict from the logging record"""
        message = {
            '@version': 1,
            'message': record.getMessage(),

            # Extra Fields
            'levelname': record.levelname,
            'logger': record.name,
            'type': self.message_type,
            'path': record.pathname,

            # required for logstash forwarder to work
            'host': self.host,
            "line": record.lineno,
            "offset": 0, # this is completely meaningless, but required
        }

        # Add extra fields
        message.update(self.get_extra_fields(record))

        # If exception, add debug info
        if record.exc_info:
            message.update(self.get_debug_fields(record))

        return message



stashward/stashward.py
"""
This implements a logstash-forwarder compatible formatter and log handler.
"""
from __future__ import print_function
import traceback
import struct
import logging
import errno
import socket
import sys
from logging.handlers import SocketHandler
from ssl import wrap_socket, CERT_REQUIRED, SSLError
try:
    from ssl import match_hostname, CertificateError
except ImportError:
    from backports.ssl_match_hostname import match_hostname, CertificateError


class ErrorCode:
    """
    We use some random numbers for error codes to make unit testing a little
    easier, and for greppability
    """
    GENERIC_SSL_FAILURE = 42218
    CN_MISMATCH = 34624
    FILE_NOT_FOUND = 17253


class StashwardFormatter(logging.Formatter):
    def __init__(self, message_type="django"):
        self.host = socket.gethostname()
        self.message_type = message_type

    def get_extra_fields(self, record):
        """Tack on all the fields from the extra parameter on logging calls"""
        # The list contains all the attributes listed in
        # http://docs.python.org/library/logging.html#logrecord-attributes
        skip_list = set([
            'args', 'asctime', 'created', 'exc_info', 'exc_text', 'filename',
            'funcName', 'id', 'levelname', 'levelno', 'lineno', 'module',
            'msecs', 'msecs', 'message', 'msg', 'name', 'pathname', 'process',
            'processName', 'relativeCreated', 'thread', 'threadName', 'extra', 'stack_info'])

        if sys.version_info < (3, 0):
            easy_types = (basestring, bool, dict, float, int, list, type(None))
        else:
            easy_types = (str, bool, dict, float, int, list, type(None))

        fields = {}

        for key, value in record.__dict__.items():
            if key not in skip_list:
                if isinstance(value, easy_types):
                    fields[key] = value
                else:
                    fields[key] = repr(value)

        return fields

    def get_debug_fields(self, record):
        """Format the exception info"""
        fields = {
            'exc_info': self.formatException(record.exc_info),
            'process': record.process,
        }

        # funcName was added in 2.5
        if not getattr(record, 'funcName', None):
            fields['funcName'] = record.funcName

        # processName was added in 2.6
        if not getattr(record, 'processName', None):
            fields['processName'] = record.processName

        return fields

    def format(self, record):
        """Create a dict from the logging record"""
        message = {
            '@version': 1,
            'message': record.getMessage(),

            # Extra Fields
            'levelname': record.levelname,
            'logger': record.name,
            'type': self.message_type,
            'path': record.pathname,

            # required for logstash forwarder to work
            'host': self.host,
            "line": record.lineno,
            "offset": 0, # this is completely meaningless, but required
        }

        # Add extra fields
        message.update(self.get_extra_fields(record))

        # If exception, add debug info
        if record.exc_info:
            message.update(self.get_debug_fields(record))

        return message


class StashwardHandler(SocketHandler):
    # The sequence number is 32 bits, so it should rollover to 0 when it gets there
    MAX_SEQUENCE_NUMBER = 2**32-1

    def __init__(self, host, port, ca_certs, message_type="django"):
        """
        Specify the host, port, and file path to the concatenated list of CAs
        """
        SocketHandler.__init__(self, host, port)
        self.formatter = StashwardFormatter(message_type)
        self.ca_certs = ca_certs
        self.sequence_number = 0

        # immediately send a packet setting the windows size to something huge
        # since we don't care about ACKs
        self.send(struct.pack("!ssI", b"1", b"W", self.MAX_SEQUENCE_NUMBER))

    def makeSocket(self):
        """Make the socket and wrap it with SSL. A valid certificate is required from a trusted CA"""
        s = SocketHandler.makeSocket(self)

        try:
            s = wrap_socket(s, cert_reqs=CERT_REQUIRED, ca_certs=self.ca_certs)
        except SSLError as e:
            # the ssl certificate was probably bad
            # since logging doesn't work, print to stderr
            traceback.print_exc()
            print("ERROR %d: This is likely caused by a CN mismatch, an invalid certificate file at %s, or a verification failure based on your CA file" % (ErrorCode.GENERIC_SSL_FAILURE, self.ca_certs), file=sys.stderr)
            s.close()
            raise
        except OSError as e:
            s.close()
            # if we get an ENOENT, that's because the ca_file is bad
            if e.errno == errno.ENOENT:
                traceback.print_exc()
                print("ERROR %d: This error is probably caused by a bad CA file at %s" % (ErrorCode.FILE_NOT_FOUND, self.ca_certs), file=sys.stderr)
            s.close()
            raise

        # some python versions do not do a CN match check, so we have to do it
        try:
            match_hostname(s.getpeercert(), self.host)
        except CertificateError as e:
            traceback.print_exc()
            print("ERROR %d: The CN provided by the server didn't match the host you connected to" % ErrorCode.CN_MISMATCH, file=sys.stderr)
            s.close()
            # raising a socket error will trigger createSocket to retry after a period of time
            raise socket.error("CN Mismatch")

        return s

    def makePickle(self, record):
        """Format the record, and turn it into a logstash compatible packet"""
        packet = self.packetize(self.formatter.format(record), self.sequence_number)
        # critical section here for incremeting the sequence_number. Notice, we
        # don't have to wrap a lock around this, because the base
        # logging.Handler will acquire a lock before calling emit(), which
        # calls this method
        self.sequence_number = self.sequence_number + 1 if self.sequence_number < self.MAX_SEQUENCE_NUMBER else 0
        return packet

    def packetize(self, data, sequence_number):
        """
        Turn a dict into a byte stream that is compatible with logstash-forwarder's protocol

        The packet looks like this:

        Header "1D" - the version and frame type (data) in ascii
        32 bit unsigned, big-endian, integer sequence number
        32 bit unsigned, big-endian, integer count of the number of (key, value) pairs
        For each key, value pair:
            32 bit unsigned, big-endian, integer number for the key length
            the key (raw bytes)
            32 bit unsigned, big-endian, integer number for the value length
            the value (raw bytes)


        The data *MUST* include a key/value pair for "host", "line" and
        "offset". That fact isn't documented anywhere but here. You're welcome.
        """
        format_string = ["!ssII"]
        string = [b"1", b"D", sequence_number, len(data)]

        for key, value in data.items():
            # encode the keys and values as utf8
            key = str(key).encode("utf8")
            value = str(value).encode("utf8")

            # tack on the key length
            format_string.append("I")
            string.append(len(key))
            # tack on the key itself
            format_string.append(str(len(key)) + "s")
            string.append(key)
            # tack on the value length
            format_string.append("I")
            string.append(len(value))
            # tack on the value
            format_string.append(str(len(value)) + "s")
            string.append(value)

        return struct.pack("".join(format_string), *string)
